File: traffic_ops/app/bin/checks/ToPingCheck.pl
#!/usr/bin/perl
#
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Plugin for the "ping" and "MTU" check.
#
# example cron entry
# 0 * * * * root /opt/traffic_ops/app/bin/checks/ToPingCheck.pl -c "{\"base_url\": \"https://localhost\", \"check_name\": \"10G\", \"name\": \"IPv4 Ping\", \"select\": \"ipAddress\"}" >> /var/log/traffic_ops/extensionCheck.log 2>&1
#
# example cron entry with select array
# 0 * * * * root /opt/traffic_ops/app/bin/checks/ToPingCheck.pl -c "{\"base_url\": \"https://localhost\", \"check_name\": \"10G\", \"name\": \"IPv4 Hostname Ping\", \"select\": [\"hostName\",\"domainName\"]}" >> /var/log/traffic_ops/extensionCheck.log 2>&1
#
# example cron entry with syslog
# 0 * * * * root /opt/traffic_ops/app/bin/checks/ToPingCheck.pl -c "{\"base_url\": \"https://localhost\", \"check_name\": \"10G\", \"name\": \"IPv4 ping\", \"select\": \"ipAddress\", \"syslog_facility\": \"local0\"}" > /dev/null 2>&1
#
# example cron entry for MTU
# 0 0 * * * root /opt/traffic_ops/app/bin/checks/ToPingCheck.pl -c "{\"base_url\": \"https://localhost\", \"check_name\": \"MTU\", \"name\": \"Max Trans Unit\", \"select\": \"ipAddress\", \"syslog_facility\": \"local0\"}" > /dev/null 2>&1
# 0 0 * * * root /opt/traffic_ops/app/bin/checks/ToPingCheck.pl -c "{\"base_url\": \"https://localhost\", \"check_name\": \"MTU\", \"name\": \"Max Trans Unit\", \"select\": \"ip6Address\", \"syslog_facility\": \"local0\"}" > /dev/null 2>&1

use strict;
use warnings;

$|++;

use Data::Dumper;
use Getopt::Std;
use Log::Log4perl qw(:easy);
use JSON;
use Extensions::Helper;
use Sys::Syslog qw(:standard :macros);

my $VERSION = "0.03";

my %args = ();
getopts( "c:f:hl:q", \%args );

if ($args{h}) {
   &help();
   exit();
}

Log::Log4perl->easy_init($ERROR);
if ( defined( $args{l} ) ) {
	if    ( $args{l} == 1 ) { Log::Log4perl->easy_init($INFO); }
	elsif ( $args{l} == 2 ) { Log::Log4perl->easy_init($DEBUG); }
	elsif ( $args{l} >= 3 ) { Log::Log4perl->easy_init($TRACE); }
   else                    { Log::Log4perl->easy_init($INFO); }
}

# For syslog messages
setlogmask(LOG_UPTO(LOG_INFO));

DEBUG( "Including DEBUG messages in output. Config is \'" . $args{c} . "\'" );
TRACE( "Including TRACE messages in output. Config is \'" . $args{c} . "\'" );

if ( !defined( $args{c} ) ) {
   ERROR "-c not defined";
   print "\n\n";
	&help();
	exit(1);
}

my $jconf = undef;
eval { $jconf = decode_json( $args{c} ) };
if ($@) {
	ERROR("Bad json config: $@");
   print "\n\n";
   &help();
	exit(1);
}

my $sslg = undef;
my $check_lng_name;
if (defined($jconf->{syslog_facility})) {
   $check_lng_name = &trim($jconf->{name});
   openlog ('ToChecks', '', $jconf->{syslog_facility});
   $sslg = 1;
}

my $force = 0;
if (defined($args{f})) {
   $force = $args{f};
}

my $quiet;
if ($args{q}) {
   $quiet = 1;
}

TRACE Dumper($jconf);
my $b_url = $jconf->{base_url};
Extensions::Helper->import();
my $ext = Extensions::Helper->new( { base_url => $b_url, token => '91504CE6-8E4A-46B2-9F9F-FE7C15228498' } );

my $jdataserver    = $ext->get(Extensions::Helper::SERVERLIST_PATH);
my $select         = $jconf->{select};
my $check_name     = &trim($jconf->{check_name});

foreach my $server ( @{$jdataserver} ) {
	if ( $server->{type} =~ m/^EDGE/ || $server->{type} =~ m/^MID/ ) {
      my $srv_nm = $server->{hostName}.".".$server->{domainName};
      my $srv_status = &trim($server->{status});
		my $ip = undef;
      my $pingable = undef;
      my ($size, $msg);

      if ($check_name =~ m/^MTU$/) {
         $size = &trim($server->{interfaceMtu});
         $size = $size - 28;
      } else {
         $size = 30;
      }

      # select in the jconf is mandatory. TODO should probably error if not there
      if ( ref($select) eq 'ARRAY' ) {
         DEBUG "select is an array";
         $select->[0] = &trim($select->[0]);
         $select->[1] = &trim($select->[1]);
         $ip = $server->{ $select->[0] } . "." . $server->{ $select->[1] };
      }
      else {
         DEBUG "select is not an array";
         $select = &trim($select);
         $ip = &trim($server->{$select});
         DEBUG "ip: ".$ip;
      }
      if (!defined($ip) || ($ip eq '')) {
         next;
      }

      if ($force == 0) {
         $pingable = &ping_check($ip, $size);
      } elsif ($force == 1) {
         $pingable = 1;
         $msg = "Force: OK";
      } elsif ($force == 2) {
         $pingable = 0;
         $msg = "Force: FAIL";
      }
      if ($pingable && $sslg) {
         $ip =~ s/\/\d+$// if ( $ip =~ /:/ );
         my @tmp = ($srv_nm,$check_name,$check_lng_name,'OK',$srv_status,$ip,
                    $server->{ipAddress},$size,$msg);
         syslog(LOG_INFO, "hostname=%s check=%s name=\"%s\" result=%s status=%s target=%s ip=%s size=%d msg=\"%s\"", @tmp);
      } elsif ($sslg) {
         $ip =~ s/\/\d+$// if ( $ip =~ /:/ );
         my @tmp = ($srv_nm,$check_name,$check_lng_name,'FAIL',$srv_status,$ip,
                    $server->{ipAddress},$size,$msg);
         syslog(LOG_ERR, "hostname=%s check=%s name=\"%s\" result=%s status=%s target=%s ip=%s size=%d msg=\"%s\"", @tmp);
      }
		DEBUG $check_name . " >> " . $server->{hostName} . ": " . $select . " = " . $ip . " ---> " . $pingable . "\n";
		$ext->post_result( $server->{id}, $check_name, $pingable ) if (!$quiet);
	}
}

closelog();

sub ping_check {
	my $ping_target = shift;    # use address to bypass DNS and FQDN to check DNS
	my $size        = shift;

	if ( !defined($ping_target) ) {
		print "Nothing to ping!\n";
		return 0;
	}

	if ( !defined($size) ) {
		$size = 30;
	}

	TRACE "Ping checking " . $ping_target." with: ".$size;

	my $cmd;
	if ( $ping_target =~ /:/ ) {
		$ping_target =~ s/\/\d+$//;
		$cmd = '/bin/ping6 -M do -s ' . $size . ' -c 2 ' . $ping_target . ' 2>&1 > /dev/null';
	}
	else {
		$cmd = '/bin/ping -M do -s ' . $size . ' -c 2 ' . $ping_target . ' 2>&1 > /dev/null';
	}

	system($cmd);
	if ( $? != 0 ) {
		ERROR $ping_target . " is NOT Pingable (with " . $size . " packet size)";
		return 0;
	}
	return 1;
}

sub ltrim { my $s = shift; $s =~ s/^\s+//;       return $s };
sub rtrim { my $s = shift; $s =~ s/\s+$//;       return $s };
sub  trim { my $s = shift; $s =~ s/^\s+|\s+$//g; return $s };

sub help() {
   print "ToPingCheck.pl -c \"{\\\"base_url\\\": \\\"https://localhost\\\", \\\"check_name\\\": \\\"10G\\\", \\\"select\\\": \\\"ipAddress\\\"[, \\\"name\\\": \\\"IPv4 Ping\\\", \\\"syslog_facility\\\": \\\"local0\\\"}]\" [-f <1-2>] [-l <1-3>]\n";
   print "\n";
   print "-c   json formatted list of variables\n";
   print "     base_url: required\n";
   print "        URL of the Traffic Ops server.\n";
   print "     check_name: required\n";
   print "        10G - IPv4 ping\n";
   print "        10G6 - IPv6 ping\n";
   print "        ILO - Lights out management interface pring.\n";
   print "        MTU - Ping with the size set to MTU value configured in Traffic Ops.\n";
   print "     select: required\n";
   print "        Examples:\n";
   print "           \"select:\" \"ipAddress\"\n";
   print "           \"select:\" [\"hostName\",\"domainName\"]\n";
   print "           \"select:\" \"ip6Address\"\n";
   print "           \"select:\" \"iloIpAddress\"\n";
   print "     name: optional\n";
   print "        The long name of this check. used in conjuction with syslog_facility.\n";
   print "     syslog_facility: optional\n";
   print "        The syslog facility to send messages. Requires the \"name\" option to\n";
   print "        be set.\n";
   print "-f   Force a FAIL or OK message\n";
   print "        1: FAIL\n";
   print "        2: OK\n";
   print "-h   Print this message\n";
   print "-l   Debug level\n";
   print "-q   Don't post results to Traffic Ops.\n";
   print "================================================================================\n";
   # the above line of equal signs is 80 columns
   print "\n";
}

File: traffic_ops/app/bin/checks/ToPingCheck.go
/*
   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at
   http://www.apache.org/licenses/LICENSE-2.0
   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

/* ToPingCheck.go
   Used for checking ILO ping,  MTU test, 10G (IPv4), and 10G6 (IPv6) pings.
*/

package main

import (
	"encoding/json"
	"flag"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/apache/trafficcontrol/lib/go-log"
	tc "github.com/apache/trafficcontrol/lib/go-tc"
	toclient "github.com/apache/trafficcontrol/traffic_ops/v3-client"
)

// Traffic Ops connection params
const AllowInsecureConnections = false
const UserAgent = "go/tc-ping-monitor"
const UseClientCache = false
const TrafficOpsRequestTimeout = time.Second * time.Duration(10)

var confForce *bool

type Config struct {
	URL    string `json:"to_url"`
	User   string `json:"user"`
	Passwd string `json:"passwd"`
}

type Server struct {
	id        int
	cdn       string
	name      string
	fqdn      string
	ilo       string
	ip4       string
	ip6       string
	mtu       int
	failcount int
	status    string
	failaddr  string
}

func NewServer(id int, name string, status string, f int) Server {
	server := Server{}
	server.id = id
	server.name = name
	server.status = status
	server.failcount = f
	return server
}

func LoadConfig(file string) (Config, error) {
	var config Config
	configFile, err := os.Open(file)
	defer configFile.Close()
	if err != nil {
		return config, err
	}
	jsonParser := json.NewDecoder(configFile)
	jsonParser.Decode(&config)
	return config, err
}

func (s *Server) ping(name string) bool {
	// I haven't found a good way to do this natively with go yet, which makes me sad
	s.failcount = 0
	size := 30 //default
	ok := true
	var addr string
	switch name {
	case "IPv4", "10G":
		log.Infof("IPv4")
		if s.ip4 != "" {
			ok, addr = ping4(size, s.ip4)
			if ok == false {
				s.failaddr = addr
			}
		}
	case "IPv6", "10G6":
		log.Infof("IPv6")
		if s.ip6 != "" {
			ok, addr = ping6(size, s.ip6)
			if ok == false {
				s.failaddr = addr
			}
		}
	case "ILO":
		log.Infof("ILO")
		if s.ilo != "" {
			match, err := regexp.MatchString(":", s.ilo)
			if err != nil {
				log.Errorf("Match error:", err)
				os.Exit(1)
			}
			if match {
				ok, addr = ping6(size, s.ilo)
			} else {
				ok, addr = ping4(size, s.ilo)
				if ok == false {
					s.failaddr = addr
				}
			}
		}
	case "MTU":
		log.Infof("MTU")
		ok4 := true
		if s.ip4 != "" {
			// subtract protocol headers from MTU to get payload size
			size = s.mtu - 28
			ok4, addr = ping4(size, s.ip4)
		} else {
			log.Warnf("no IPv4 address detected (skipping)")
		}
		if ok4 == false {
			s.failaddr = addr
		}
		ok6 := true
		if s.ip6 != "" {
			// subtract protocol headers from MTU to get payload size
			size = s.mtu - 48
			ok6, addr = ping6(size, s.ip6)
		} else {
			log.Warnf("no IPv6 address detected (skipping)")
		}
		if ok6 == false {
			if len(s.failaddr) > 0 {
				s.failaddr = s.failaddr + "," + addr
			} else {
				s.failaddr = addr
			}
		}
		if !(ok4 && ok6) {
			ok = false
		}
	}
	return ok
}

func ping4(size int, addr string) (bool, string) {
	log.Infof("size: ", strconv.Itoa(size))
	out, err := exec.Command("/bin/ping", "-M", "do", "-s", strconv.Itoa(size), "-c", "2", addr).Output()
	if err != nil {
		log.Warnf("ping failed for %s: %s", addr, err.Error())
		return false, addr
	}
	log.Debugf("ping output:\n%v", out)
	return true, addr
}

func ping6(size int, addr string) (bool, string) {
	log.Infof("size: ", strconv.Itoa(size))
	out, err := exec.Command("/bin/ping6", "-M", "do", "-s", strconv.Itoa(size), "-c", "2", addr).Output()
	if err != nil {
		log.Warnf("ping failed for %s: %s", addr, err.Error())
		return false, addr
	}
	log.Debugf("ping output:\n%v", out)
	return true, addr
}

func main() {
	var cpath_new string
	var ok bool

	jobStart := time.Now()

	// define default config file path
	cpath, err := filepath.Abs(filepath.Dir(os.Args[0]))
	if err != nil {
		log.Errorf("Config error:", err)
		os.Exit(1)
	}
	cpath_new = strings.Replace(cpath, "/bin/checks", "/conf/check-config.json", 1)

	// command-line flags
	confPtr := flag.String("conf", cpath_new, "Config file path")
	confName := flag.String("name", "undef", "'10G|IPv4', '10G6|IPv6', 'ILO', 'MTU'")
	confInclude := flag.String("host", "undef", "Specific host or regex to include (optional)")
	confCdn := flag.String("cdn", "all", "Check specific CDN by name")
	confExclude := flag.String("exclude", "undef", "Hostname regex to exclude")
	confReset := flag.Bool("reset", false, "Reset check values in TO to 'blank' state")
	confQuiet := flag.Bool("q", false, "Do not send updates to TO")
	confForce = flag.Bool("f", false, "Force a failure result")
	flag.Parse()

	reName, err := regexp.Compile("^(10G|10G6|IPv4|IPv6|ILO|MTU)$")
	if err != nil {
		log.Errorf("supplied exclusion regex does not compile:", err)
		os.Exit(1)
	}
	if !(reName.Match([]byte(*confName))) {
		log.Errorf("Check name must be one of the following:")
		log.Errorf("'10G' (legacy) or 'IPv4' (new) for IPv4 interface check")
		log.Errorf("'10G6' (legacy) or 'IPv6' (new) for IPv6 interface check")
		log.Errorf("'ILO' out-of-band mgmt interface check")
		log.Errorf("'MTU' uses the MTU value for the server in TO to check MTU (checks both v4 and v6, if available)")
		os.Exit(1)
	}

	// load config json
	config, err := LoadConfig(*confPtr)
	if err != nil {
		log.Errorf("Error loading config:", err)
		os.Exit(1)
	}

	// connect to TO
	session, _, err := toclient.LoginWithAgent(
		config.URL,
		config.User,
		config.Passwd,
		AllowInsecureConnections,
		UserAgent,
		UseClientCache,
		TrafficOpsRequestTimeout)
	if err != nil {
		log.Errorf("An error occurred while logging in: %v\n", err)
		os.Exit(1)
	}

	// Make TO API call for server details
	var servers tc.ServersV3Response
	servers, _, err = session.GetServersWithHdr(nil, nil)
	if err != nil {
		log.Errorf("An error occurred while getting servers: %v\n", err)
		os.Exit(1)
	}

	for _, server := range servers.Response {
		re, err := regexp.Compile("^(MID|EDGE).*")
		if err != nil {
			log.Errorf("supplied exclusion regex does not compile:", err)
			os.Exit(1)
		}
		if re.Match([]byte(server.Type)) {
			serverStart := time.Now()
			if *confInclude != "undef" {
				re_inc, err := regexp.Compile(*confInclude)
				if err != nil {
					log.Errorf("supplied exclusion regex does not compile:", err)
					os.Exit(1)
				}
				if !re_inc.MatchString(*server.HostName) {
					log.Debugf("%s does not match the provided include regex, skipping", server.HostName)
					continue
				}
			}
			if *confCdn != "all" && *confCdn != *server.CDNName {
				log.Debugf("%s is not assinged to the specified CDN '%s', skipping", server.HostName, *confCdn)
				continue
			}
			if *confExclude != "undef" {
				re, err := regexp.Compile(*confExclude)
				if err != nil {
					log.Errorf("supplied exclusion regex does not compile:", err)
					os.Exit(1)
				}
				if re.MatchString(*server.HostName) {
					log.Debugf("%s matches the provided exclude regex, skipping", server.HostName)
					continue
				}
			}
			s := NewServer(*server.ID, *server.HostName, *server.Status, -1)
			defaulStatusValue := -1
			var statusData tc.ServercheckRequestNullable
			statusData.ID = &s.id
			statusData.Name = confName
			statusData.HostName = &s.name
			statusData.Value = &defaulStatusValue
			s.fqdn = s.name + "." + *server.DomainName
			for _, interf := range server.Interfaces {
				for _, addr := range interf.IPAddresses {
					if s.ip4 == "" && strings.Count(addr.Address, ":") == 0 {
						s.ip4 = strings.Split(addr.Address, "/")[0]
					}
					if s.ip6 == "" && strings.Count(addr.Address, ":") > 0 {
						s.ip6 = strings.Split(addr.Address, "/")[0]
					}
				}
			}
			if server.ILOIPAddress == nil {
				s.ilo = ""
			} else {
				if strings.Contains(*server.ILOIPAddress, "/") {
					s.ilo = strings.Split(*server.ILOIPAddress, "/")[0]
				} else {
					s.ilo = *server.ILOIPAddress
				}
			}
			s.cdn = *server.CDNName
			s.mtu = int(*server.Interfaces[0].MTU)
			log.Infof("Next server=%s status=%s", s.fqdn, s.status)
			if (s.status == "REPORTED" || s.status == "ADMIN_DOWN") && *confReset != true {
				ok = s.ping(*confName)
				log.Infof("ok: %v", ok)
				if ok == false {
					s.failcount = 1
				}
			}

			// send status update to TO
			if s.failcount == -1 {
				// server not checked
				*statusData.Value = -1
			} else if s.failcount > 0 {
				// server had failures
				log.Infof("result=failure server=%s status=%s check=%s addr=%s", s.fqdn, s.status, *confName, s.failaddr)
				*statusData.Value = 0
			} else {
				// server looks OK
				log.Infof("result=success server=%s status=%s", s.fqdn, s.status)
				*statusData.Value = 1
			}
			serverElapsed := time.Since(serverStart)
			log.Infof("Finished checking server=%s result=%d cdn=%s elapsed=%s", s.fqdn, *statusData.Value, s.cdn, serverElapsed)
			if *confQuiet == false {
				log.Debugf("Sending update to TO")
				_, _, err := session.InsertServerCheckStatus(statusData)
				if err != nil {
					log.Errorf("Error updating server check status with TO:", err)
				}
			} else {
				log.Debugf("Skipping update to TO")
			}
		}
	}
	jobElapsed := time.Since(jobStart)
	log.Infof("Job complete totaltime=%s", jobElapsed)
	os.Exit(0)
}